File: src/main/kotlin/domain/auth/UserActions.kt
@file:OptIn(ExperimentalTime::class)

package com.kitledger.domain.auth

import com.kitledger.services.database.SystemPermissionsTable
import com.kitledger.services.database.UsersTable
import com.kitledger.services.utils.generateUuidV7
import org.jetbrains.exposed.v1.r2dbc.insert
import org.jetbrains.exposed.v1.r2dbc.transactions.suspendTransaction
import kotlin.random.Random
import kotlin.time.Clock
import kotlin.time.ExperimentalTime


/**
 * Creates a super user (System or Root User) with complete system permissions.
 * @param firstName the user's first name.
 * @param lastName the user's last name.
 * @param email the user's email.
 * @return a result type with the NewSuperUser object on success, or an exception on failure.
 */
suspend fun createSuperUser(firstName: String, lastName: String, email: String): Result<NewSuperUser> {

    try {
        val randomPassword = generateRandomPassword()

        if (randomPassword.isEmpty()) {
            throw RuntimeException("Failed to generate random password")
        }

        val hashedPassword = hashPassword(randomPassword)
        val userId = generateUuidV7()

        val superUser = suspendTransaction {
            val userResult = UsersTable.insert {
                it[UsersTable.id] = userId
                it[UsersTable.firstName] = firstName
                it[UsersTable.lastName] = lastName
                it[UsersTable.email] = email
                it[UsersTable.passwordHash] = hashedPassword
                it[UsersTable.createdAt] = Clock.System.now()
                it[UsersTable.updatedAt] = Clock.System.now()
            }

            val user = userResult.resultedValues?.get(0)?.toUser()
                ?: throw IllegalStateException("Failed to insert user")

            SystemPermissionsTable.insert {
                it[SystemPermissionsTable.id] = generateUuidV7()
                it[SystemPermissionsTable.permission] = SystemPermissionEnum.KL_SYSTEM_ADMIN
                it[SystemPermissionsTable.userId] = user.id
                it[SystemPermissionsTable.createdAt] = Clock.System.now()
                it[SystemPermissionsTable.updatedAt] = Clock.System.now()
            }

            val apiToken = createToken(userId) ?: throw IllegalStateException("Failed to create API token")
            val tokenId = apiToken.id

            val signedToken = signToken(TokenType.API, tokenId) ?: throw IllegalStateException("Failed to sign token")

            NewSuperUser(
                userId,
                firstName,
                lastName,
                email,
                randomPassword,
                signedToken
            )
        }

        return Result.success(superUser)
    } catch (e: Exception) {
        return Result.failure(e)
    }
}

/**
 * Generates a random password.
 * @param length the length of the password.
 * @return the password as a string.
 */
private fun generateRandomPassword(length: Int = 20): String {
    val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9') + "!@#$%^&*()_+-=".toList()

    val passwordBuilder = StringBuilder(length)

    for (i in 0 until length) {
        passwordBuilder.append(charPool[Random.nextInt(charPool.size)])
    }

    return passwordBuilder.toString()
}

File: src/domain/auth/user_actions.ts
import { db } from "../../services/database/db.js";
import { randomBytes } from "node:crypto";
import {v7} from "uuid";
import { SYSTEM_ADMIN_PERMISSION } from "./permission_actions.js";
import { createToken } from "./token_actions.js";
import { assembleApiTokenJwtPayload, signToken } from "./jwt_actions.js";
import { workerPool } from "../../services/workers/pool.js";
import { availableWorkerTasks } from "../../services/workers/worker.js";
import { type User } from "./types.js";
import { system_permissions, users } from "../../services/database/schema.js";

export type NewSuperUser = Pick<User, "id" | "first_name" | "last_name" | "email"> & {
	password: string;
	api_token: string;
};

export async function createSuperUser(
	firstName: string,
	lastName: string,
	email: string,
): Promise<NewSuperUser | null> {
	const newSuperUser: NewSuperUser | null = await db.transaction(async (tx) => {
		try {
			/**
			 * generate a random password.
			 */
			const password = randomBytes(20).toString("hex");
			let passwordHash = null;

			try {
				passwordHash = await workerPool?.execute(password, availableWorkerTasks.HASH_PASSWORD);
			}
			catch (error) {
				console.error("Error hashing password:", error);
				passwordHash = null;
			}

			if (!passwordHash) {
				throw new Error("Failed to hash password");
			}

			const userId = v7();
			const newUser = await tx
				.insert(users)
				.values({
					id: userId,
					first_name: firstName,
					last_name: lastName,
					email: email,
					password_hash: passwordHash as string,
					created_at: new Date(),
				})
				.returning();

			if (!newUser[0]) {
				throw new Error("Failed to create user");
			}

			await tx.insert(system_permissions).values({
				id: v7(),
				user_id: newUser[0].id,
				permission: SYSTEM_ADMIN_PERMISSION,
			});

			return {
				id: newUser[0].id,
				first_name: firstName,
				last_name: lastName,
				email: email,
				password: password,
				api_token: "",
			};
		}
		catch (error) {
			console.error("Error creating super user:", error);
			tx.rollback();
			return null;
		}
	});

	// Create API token for the new super user, using the encapsulated function.

	if (!newSuperUser) {
		return null;
	}

	const tokenName = `${firstName} ${lastName} Super User Token`.slice(0, 64);
	const apiToken = await createToken(newSuperUser.id, tokenName);

	if (!apiToken) {
		console.error("Failed to create API token for super user");
	}

	newSuperUser.api_token = await signToken(assembleApiTokenJwtPayload(apiToken));

	return newSuperUser;
}
